README
DiVine2
=======

DiVine2 (DirectFB Virtual input extension) consists of:
  - the DiVine input driver which reads raw input events from a pipe and dispatches them,
  - the DiVine library which handles the connection to the input driver and provides an
    interface for sending virtual events.


ChangeLog
commit 6549ca6691b9fb865257c599fdb343d24848914e
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Wed Feb 1 09:29:45 2023 +0100

    DirectFB Virtual input extension

commit 092ad29d180a713b688b8b8dd72ad87355ddbf6c
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Wed Feb 1 09:16:08 2023 +0100

    Initial revision


src/divine.c
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#include <direct/direct.h>
#include <idivine.h>
#include <misc/conf.h>

D_DEBUG_DOMAIN( DiVine_Main, "DiVine/Main", "DiVine Main Functions" );

/**********************************************************************************************************************/

DFBResult
DiVineInit( int   *argc,
            char **argv[] )
{
     DFBResult ret;

     ret = dfb_config_init( argc, argv );
     if (ret)
          return ret;

     return DFB_OK;
}

DFBResult
DiVineCreate( IDiVine **ret_interface )
{
     DFBResult  ret;
     IDiVine   *divine;

     D_DEBUG_AT( DiVine_Main, "%s( %p )\n", __FUNCTION__, ret_interface );

     if (!dfb_config) {
          D_ERROR( "DiVine/Main: DiVineInit() has to be called before DiVineCreate()!\n" );
          return DFB_INIT;
     }

     if (!ret_interface)
          return DFB_INVARG;

     direct_initialize();

     DIRECT_ALLOCATE_INTERFACE( divine, IDiVine );

     ret = IDiVine_Construct( divine );
     if (ret)
          return ret;

     *ret_interface = divine;

     return DFB_OK;
}


src/idivine.h
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#ifndef __IDIVINE_H__
#define __IDIVINE_H__

#include <coretypes_divine.h>

/*
 * initializes interface struct and private data
 */
DFBResult IDiVine_Construct( IDiVine *thiz );

#endif


include/divine.h
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#ifndef __DIVINE_H__
#define __DIVINE_H__

#ifdef __cplusplus
extern "C" {
#endif

#include <directfb.h>

/*
 * Main interface of DiVine.
 */
D_DECLARE_INTERFACE( IDiVine )

/**********************************************************************************************************************/

#define DIVINE_API

/**********************************************************************************************************************/

/*
 * Parses the command-line and initializes some variables.
 * You absolutely need to call this before doing anything else.
 * Removes all options used by DiVine from argv.
 */
DFBResult DIVINE_API DiVineInit(
     int                                    *argc,               /* pointer to main()'s argc */
     char                                  **argv[]              /* pointer to main()'s argv */
);

/*
 * Creates the main interface.
 */
DFBResult DIVINE_API DiVineCreate(
     IDiVine                               **ret_interface       /* pointer to the created interface */
);

/**********************************************************************************************************************/

/***********
 * IDiVine *
 ***********/

/*
 * IDiVine is the main interface. It can be retrieved by a
 * call to DiVineCreate().
 */
D_DEFINE_INTERFACE( IDiVine,

   /** Events **/

     /*
      * Send an input event.
      */
     DFBResult (*SendEvent) (
          IDiVine                           *thiz,
          const DFBInputEvent               *event
     );

     /*
      * Send an input key symbol.
      */
     DFBResult (*SendSymbol) (
          IDiVine                           *thiz,
          DFBInputDeviceKeySymbol            symbol
     );
);

#ifdef __cplusplus
}
#endif

#endif


src/core_divine.c
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#include <core_divine.h>

D_DEBUG_DOMAIN( CoreDiVine_Main, "CoreDiVine/Main", "DiVine Core" );

/**********************************************************************************************************************/

struct __DV_CoreDiVine {
     int fd; /* file descriptor of the connection (pipe) */
};

/**********************************************************************************************************************/

CoreDiVine *
divine_open( const char *path )
{
     int         fd;
     CoreDiVine *divine;

     D_DEBUG_AT( CoreDiVine_Main, "%s()\n", __FUNCTION__ );

     /* Open pipe. */
     fd = open( path, O_WRONLY );
     if (fd < 0) {
          D_PERROR( "CoreDiVine: Could not open pipe '%s'!\n", path );
          return NULL;
     }

     /* Allocate connection object. */
     divine = D_CALLOC( 1, sizeof(CoreDiVine) );
     if (!divine) {
          D_OOM();
          return NULL;
     }

     /* Fill out connection information. */
     divine->fd = fd;

     return divine;
}

void
divine_send_event( CoreDiVine          *divine,
                   const DFBInputEvent *event )
{
     D_DEBUG_AT( CoreDiVine_Main, "%s()\n", __FUNCTION__ );

     /* Write event to pipe. */
     write( divine->fd, event, sizeof(DFBInputEvent) );
}

void
divine_send_symbol( CoreDiVine              *divine,
                    DFBInputDeviceKeySymbol  symbol )
{
     DFBInputEvent event;

     D_DEBUG_AT( CoreDiVine_Main, "%s()\n", __FUNCTION__ );

     /* Construct 'press' event. */
     event.flags      = DIEF_KEYSYMBOL;
     event.type       = DIET_KEYPRESS;
     event.key_symbol = symbol;

     /* Write 'press' event to pipe. */
     write( divine->fd, &event, sizeof(DFBInputEvent) );

     /* Turn into 'release' event. */
     event.type = DIET_KEYRELEASE;

     /* Write 'release' event to pipe. */
     write( divine->fd, &event, sizeof(DFBInputEvent) );
}

void
divine_close( CoreDiVine *divine )
{
     D_DEBUG_AT( CoreDiVine_Main, "%s()\n", __FUNCTION__ );

     /* Close pipe. */
     close( divine->fd );

     /* Free connection object. */
     D_FREE( divine );
}


src/core_divine.h
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#ifndef __CORE_DIVINE_H__
#define __CORE_DIVINE_H__

#include <coretypes_divine.h>

/**********************************************************************************************************************/

/*
 * Returns the DiVine connection object (connection to the input driver by opening the pipe specified by 'path').
 */
CoreDiVine *divine_open       ( const char              *path );

/*
 * Sends an input event.
 */
void        divine_send_event ( CoreDiVine              *divine,
                                const DFBInputEvent     *event );

/*
 * Sends a press and a release event for the specified symbol.
 */
void        divine_send_symbol( CoreDiVine              *divine,
                                DFBInputDeviceKeySymbol  symbol );

/*
 * Closes the pipe and destroys the connection object.
 */
void        divine_close      ( CoreDiVine              *divine );

#endif


src/coretypes_divine.h
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#ifndef __CORETYPES_DIVINE_H__
#define __CORETYPES_DIVINE_H__

#include <divine.h>

/**********************************************************************************************************************/

typedef struct __DV_CoreDiVine CoreDiVine;

#endif


src/idivine.c
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#include <core_divine.h>
#include <direct/direct.h>
#include <idivine.h>

D_DEBUG_DOMAIN( DiVine, "IDiVine", "IDiVine Interface" );

/**********************************************************************************************************************/

#define PIPE_PATH "/tmp/divine"

/*
 * private data struct of IDiVine
 */
typedef struct {
     int     ref;        /* reference counter */

     CoreDiVine *divine; /* the core object */
} IDiVine_data;

/**********************************************************************************************************************/

static void
IDiVine_Destruct( IDiVine *thiz )
{
     IDiVine_data *data = thiz->priv;

     D_DEBUG_AT( DiVine, "%s( %p )\n", __FUNCTION__, thiz );

     divine_close( data->divine );

     DIRECT_DEALLOCATE_INTERFACE( thiz );

     direct_shutdown();
}

static DirectResult
IDiVine_AddRef( IDiVine *thiz )
{
     DIRECT_INTERFACE_GET_DATA( IDiVine )

     D_DEBUG_AT( DiVine, "%s( %p )\n", __FUNCTION__, thiz );

     data->ref++;

     return DFB_OK;
}

static DirectResult
IDiVine_Release( IDiVine *thiz )
{
     DIRECT_INTERFACE_GET_DATA( IDiVine )

     D_DEBUG_AT( DiVine, "%s( %p )\n", __FUNCTION__, thiz );

     if (--data->ref == 0)
          IDiVine_Destruct( thiz );

     return DFB_OK;
}

static DFBResult
IDiVine_SendEvent( IDiVine             *thiz,
                   const DFBInputEvent *event )
{
     DIRECT_INTERFACE_GET_DATA( IDiVine )

     D_DEBUG_AT( DiVine, "%s( %p, %p )\n", __FUNCTION__, thiz, event );

     divine_send_event( data->divine, event );

     return DFB_OK;
}

static DFBResult
IDiVine_SendSymbol( IDiVine                 *thiz,
                    DFBInputDeviceKeySymbol  symbol )
{
     DIRECT_INTERFACE_GET_DATA( IDiVine )

     D_DEBUG_AT( DiVine, "%s( %p, 0x%08x )\n", __FUNCTION__, thiz, symbol );

     divine_send_symbol( data->divine, symbol );

     return DFB_OK;
}

DFBResult
IDiVine_Construct( IDiVine *thiz )
{
     DIRECT_ALLOCATE_INTERFACE_DATA( thiz, IDiVine )

     D_DEBUG_AT( DiVine, "%s( %p )\n", __FUNCTION__, thiz );

     data->divine = divine_open( PIPE_PATH );
     if (!data->divine) {
          DIRECT_DEALLOCATE_INTERFACE( thiz );
          return DFB_FAILURE;
     }

     data->ref = 1;

     thiz->AddRef     = IDiVine_AddRef;
     thiz->Release    = IDiVine_Release;
     thiz->SendEvent  = IDiVine_SendEvent;
     thiz->SendSymbol = IDiVine_SendSymbol;

     return DFB_OK;
}
